File: ios/MarrakechGuide/MarrakechGuide/Core/Services/HeadingService.swift
import Foundation
import CoreLocation
import os.log

/// Confidence level for compass heading readings.
public enum HeadingConfidence: String, CaseIterable, Sendable {
    /// Heading is reliable and accurate
    case good
    /// Heading may be affected by magnetic interference
    case weak
    /// Heading is not available (no sensors or significant interference)
    case unavailable
}

/// Protocol for heading/compass services.
///
/// This service wraps CLLocationManager heading updates with:
/// - Confidence indicators for heading reliability
/// - Battery-safe start/stop patterns
/// - UI update throttling (10-20Hz max)
/// - Automatic calibration detection
///
/// Usage:
/// ```swift
/// let headingService = HeadingServiceImpl.shared
/// headingService.startUpdates()
/// // ... use currentHeading and headingConfidence
/// headingService.stopUpdates()
/// ```
public protocol HeadingService: AnyObject, Sendable {
    /// Current magnetic heading in degrees (0-360, 0 = magnetic north).
    /// Nil if heading is unavailable.
    var currentHeading: CLHeading? { get }

    /// True heading adjusted for magnetic declination (if location available).
    /// Falls back to magnetic heading if true heading unavailable.
    var trueHeadingDegrees: Double? { get }

    /// Confidence level of the current heading reading.
    var headingConfidence: HeadingConfidence { get }

    /// Whether heading services are available on this device.
    var isHeadingAvailable: Bool { get }

    /// Starts receiving heading updates.
    /// Updates are throttled to preserve battery.
    /// Call stopUpdates() when compass screen is no longer visible.
    func startUpdates()

    /// Stops receiving heading updates.
    /// Always call this when the compass screen closes.
    func stopUpdates()

    /// Requests the system to display heading calibration if needed.
    /// Returns true if calibration was shown.
    @MainActor
    func requestCalibration() -> Bool
}

/// Implementation of HeadingService using CLLocationManager.
///
/// Thread-safety: All CLLocationManager operations happen on the main thread.
/// Public interface is Sendable for use from async contexts.
public final class HeadingServiceImpl: NSObject, HeadingService, @unchecked Sendable {

    // MARK: - Singleton

    public static let shared = HeadingServiceImpl()

    // MARK: - Private Properties

    private let locationManager: CLLocationManager
    private let logger = Logger(subsystem: "com.marrakechguide", category: "HeadingService")

    /// Lock for thread-safe access to mutable state
    private let lock = NSLock()

    /// Stored current heading (accessed with lock)
    private var _currentHeading: CLHeading?

    /// Stored confidence (accessed with lock)
    private var _headingConfidence: HeadingConfidence = .unavailable

    /// Whether updates are currently active
    private var isUpdating = false

    /// Timeout timer for automatic stop
    private var timeoutTimer: Timer?

    /// Timeout duration in seconds (10 minutes default)
    private let timeoutDuration: TimeInterval = 600

    /// Last time heading was updated (for throttling)
    private var lastUpdateTime: Date?

    /// Minimum interval between heading updates (50ms = 20Hz)
    private let minUpdateInterval: TimeInterval = 0.05

    // MARK: - Public Properties

    public var currentHeading: CLHeading? {
        lock.lock()
        defer { lock.unlock() }
        return _currentHeading
    }

    public var trueHeadingDegrees: Double? {
        guard let heading = currentHeading else { return nil }

        // Use true heading if available (requires location fix)
        if heading.trueHeading >= 0 {
            return heading.trueHeading
        }

        // Fall back to magnetic heading
        if heading.magneticHeading >= 0 {
            return heading.magneticHeading
        }

        return nil
    }

    public var headingConfidence: HeadingConfidence {
        lock.lock()
        defer { lock.unlock() }
        return _headingConfidence
    }

    public var isHeadingAvailable: Bool {
        CLLocationManager.headingAvailable()
    }

    // MARK: - Initialization

    private override init() {
        locationManager = CLLocationManager()
        super.init()

        locationManager.delegate = self

        // Set heading filter to reduce update frequency (1 degree change minimum)
        locationManager.headingFilter = 1.0
    }

    // MARK: - Public Methods

    public func startUpdates() {
        guard isHeadingAvailable else {
            logger.info("Heading services not available on this device")
            updateConfidence(.unavailable)
            return
        }

        guard !isUpdating else {
            logger.debug("Heading updates already active")
            return
        }

        isUpdating = true
        locationManager.startUpdatingHeading()
        logger.info("Started heading updates")

        // Set timeout to prevent indefinite battery drain
        resetTimeoutTimer()
    }

    public func stopUpdates() {
        guard isUpdating else { return }

        isUpdating = false
        locationManager.stopUpdatingHeading()
        cancelTimeoutTimer()
        logger.info("Stopped heading updates")
    }

    @MainActor
    public func requestCalibration() -> Bool {
        // CLLocationManager will show calibration UI automatically when needed
        // We can't force it, but we can check if calibration is needed
        if let heading = currentHeading, heading.headingAccuracy < 0 {
            logger.info("Heading calibration needed")
            return true
        }
        return false
    }

    // MARK: - Private Methods

    private func resetTimeoutTimer() {
        cancelTimeoutTimer()

        timeoutTimer = Timer.scheduledTimer(withTimeInterval: timeoutDuration, repeats: false) { [weak self] _ in
            self?.logger.warning("Heading updates timed out after \(self?.timeoutDuration ?? 0) seconds")
            self?.stopUpdates()
        }
    }

    private func cancelTimeoutTimer() {
        timeoutTimer?.invalidate()
        timeoutTimer = nil
    }

    private func updateHeading(_ heading: CLHeading) {
        lock.lock()
        _currentHeading = heading
        lock.unlock()
    }

    private func updateConfidence(_ confidence: HeadingConfidence) {
        lock.lock()
        _headingConfidence = confidence
        lock.unlock()
    }

    private func calculateConfidence(from accuracy: CLLocationDirection) -> HeadingConfidence {
        // headingAccuracy is in degrees, negative means invalid
        if accuracy < 0 {
            return .unavailable
        } else if accuracy <= 15 {
            return .good
        } else {
            return .weak
        }
    }

    private func shouldThrottleUpdate() -> Bool {
        guard let lastUpdate = lastUpdateTime else {
            lastUpdateTime = Date()
            return false
        }

        let elapsed = Date().timeIntervalSince(lastUpdate)
        if elapsed >= minUpdateInterval {
            lastUpdateTime = Date()
            return false
        }

        return true
    }
}

// MARK: - CLLocationManagerDelegate

extension HeadingServiceImpl: CLLocationManagerDelegate {

    public func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        // Throttle updates to 20Hz max
        if shouldThrottleUpdate() {
            return
        }

        updateHeading(newHeading)

        // Calculate confidence based on accuracy
        let confidence = calculateConfidence(from: newHeading.headingAccuracy)
        updateConfidence(confidence)

        logger.debug("Heading updated: \(newHeading.magneticHeading)° (accuracy: \(newHeading.headingAccuracy)°, confidence: \(confidence.rawValue))")

        // Reset timeout since we're getting updates
        if isUpdating {
            resetTimeoutTimer()
        }
    }

    public func locationManagerShouldDisplayHeadingCalibration(_ manager: CLLocationManager) -> Bool {
        // Allow system to show calibration UI when needed
        logger.info("Heading calibration requested by system")
        return true
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        logger.error("Heading manager error: \(error.localizedDescription)")
        updateConfidence(.unavailable)
    }
}

File: android/app/src/main/kotlin/com/marrakechguide/core/service/HeadingService.kt
package com.marrakechguide.core.service

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.util.Log
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.Timer
import java.util.TimerTask
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Confidence level for compass heading readings.
 */
enum class HeadingConfidence {
    /** Heading is reliable and accurate */
    GOOD,
    /** Heading may be affected by magnetic interference */
    WEAK,
    /** Heading is not available (no sensors or significant interference) */
    UNAVAILABLE
}

/**
 * Interface for heading/compass services.
 *
 * This service wraps SensorManager rotation vector sensor with:
 * - Confidence indicators for heading reliability
 * - Battery-safe start/stop patterns
 * - UI update throttling (10-20Hz max)
 * - Automatic calibration detection
 */
interface HeadingService {
    /** Current heading in degrees (0-360, 0 = north), null if unavailable. */
    val currentHeading: StateFlow<Float?>

    /** Confidence level of the current heading reading. */
    val headingConfidence: StateFlow<HeadingConfidence>

    /** Whether heading services are available on this device. */
    val isHeadingAvailable: Boolean

    /**
     * Starts receiving heading updates.
     * Updates are throttled to preserve battery.
     * Call stopUpdates() when compass screen is no longer visible.
     */
    fun startUpdates()

    /**
     * Stops receiving heading updates.
     * Always call this when the compass screen closes.
     */
    fun stopUpdates()

    /**
     * Requests calibration by waving device in figure-8 pattern.
     * This is typically triggered by low accuracy readings.
     *
     * @return true if calibration is needed
     */
    fun needsCalibration(): Boolean
}

/**
 * Implementation of HeadingService using SensorManager.
 *
 * Uses TYPE_ROTATION_VECTOR sensor which fuses accelerometer, magnetometer,
 * and optionally gyroscope for stable heading readings.
 *
 * This implementation:
 * - Throttles updates to 20Hz max (50ms minimum interval)
 * - Includes automatic timeout after 10 minutes
 * - Detects low accuracy and signals calibration needed
 */
@Singleton
class HeadingServiceImpl @Inject constructor(
    @ApplicationContext private val context: Context
) : HeadingService, SensorEventListener {

    companion object {
        private const val TAG = "HeadingService"
        private const val TIMEOUT_DURATION_MS = 600_000L // 10 minutes
        private const val MIN_UPDATE_INTERVAL_NS = 50_000_000L // 50ms = 20Hz
        private const val GOOD_ACCURACY_THRESHOLD = 15f // degrees
    }

    private val sensorManager: SensorManager =
        context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    private val rotationSensor: Sensor? =
        sensorManager.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR)

    private val magneticSensor: Sensor? =
        sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)

    private val _currentHeading = MutableStateFlow<Float?>(null)
    override val currentHeading: StateFlow<Float?> = _currentHeading.asStateFlow()

    private val _headingConfidence = MutableStateFlow(HeadingConfidence.UNAVAILABLE)
    override val headingConfidence: StateFlow<HeadingConfidence> = _headingConfidence.asStateFlow()

    override val isHeadingAvailable: Boolean
        get() = rotationSensor != null || magneticSensor != null

    private var isUpdating = false
    private var timeoutTimer: Timer? = null
    private var lastUpdateTimeNs = 0L

    // Rotation matrix and orientation arrays (reused to avoid allocation)
    private val rotationMatrix = FloatArray(9)
    private val orientationAngles = FloatArray(3)

    // For accelerometer/magnetometer fallback
    private val accelerometerReading = FloatArray(3)
    private val magnetometerReading = FloatArray(3)
    private var hasAccelerometer = false
    private var hasMagnetometer = false

    private val accelerometerSensor: Sensor? =
        sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER)

    override fun startUpdates() {
        if (!isHeadingAvailable) {
            Log.i(TAG, "Heading services not available on this device")
            _headingConfidence.value = HeadingConfidence.UNAVAILABLE
            return
        }

        if (isUpdating) {
            Log.d(TAG, "Heading updates already active")
            return
        }

        isUpdating = true

        // Prefer rotation vector sensor (fused, more stable)
        if (rotationSensor != null) {
            sensorManager.registerListener(
                this,
                rotationSensor,
                SensorManager.SENSOR_DELAY_UI
            )
            Log.i(TAG, "Started heading updates using rotation vector sensor")
        } else {
            // Fallback to accelerometer + magnetometer
            accelerometerSensor?.let {
                sensorManager.registerListener(
                    this,
                    it,
                    SensorManager.SENSOR_DELAY_UI
                )
            }
            magneticSensor?.let {
                sensorManager.registerListener(
                    this,
                    it,
                    SensorManager.SENSOR_DELAY_UI
                )
            }
            Log.i(TAG, "Started heading updates using accelerometer/magnetometer fallback")
        }

        // Set timeout to prevent indefinite battery drain
        resetTimeoutTimer()
    }

    override fun stopUpdates() {
        if (!isUpdating) return

        isUpdating = false
        sensorManager.unregisterListener(this)
        cancelTimeoutTimer()
        Log.i(TAG, "Stopped heading updates")
    }

    override fun needsCalibration(): Boolean {
        return _headingConfidence.value == HeadingConfidence.WEAK
    }

    // MARK: - SensorEventListener

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_ROTATION_VECTOR -> handleRotationVector(event)
            Sensor.TYPE_ACCELEROMETER -> handleAccelerometer(event)
            Sensor.TYPE_MAGNETIC_FIELD -> handleMagnetometer(event)
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
        val confidence = when (accuracy) {
            SensorManager.SENSOR_STATUS_ACCURACY_HIGH -> HeadingConfidence.GOOD
            SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM -> HeadingConfidence.GOOD
            SensorManager.SENSOR_STATUS_ACCURACY_LOW -> HeadingConfidence.WEAK
            SensorManager.SENSOR_STATUS_UNRELIABLE -> HeadingConfidence.WEAK
            else -> HeadingConfidence.UNAVAILABLE
        }

        _headingConfidence.value = confidence
        Log.d(TAG, "Sensor accuracy changed: $accuracy (confidence: $confidence)")
    }

    // MARK: - Private Methods

    private fun handleRotationVector(event: SensorEvent) {
        // Throttle updates to 20Hz
        if (shouldThrottleUpdate(event.timestamp)) return

        // Get rotation matrix from rotation vector
        SensorManager.getRotationMatrixFromVector(rotationMatrix, event.values)

        // Get orientation angles (azimuth is the first element)
        SensorManager.getOrientation(rotationMatrix, orientationAngles)

        // Convert azimuth from radians to degrees (0-360)
        val azimuthRadians = orientationAngles[0]
        var azimuthDegrees = Math.toDegrees(azimuthRadians.toDouble()).toFloat()

        // Normalize to 0-360 range
        if (azimuthDegrees < 0) {
            azimuthDegrees += 360f
        }

        updateHeading(azimuthDegrees)
    }

    private fun handleAccelerometer(event: SensorEvent) {
        System.arraycopy(event.values, 0, accelerometerReading, 0, accelerometerReading.size)
        hasAccelerometer = true
        tryComputeHeadingFromFallback(event.timestamp)
    }

    private fun handleMagnetometer(event: SensorEvent) {
        System.arraycopy(event.values, 0, magnetometerReading, 0, magnetometerReading.size)
        hasMagnetometer = true
        tryComputeHeadingFromFallback(event.timestamp)
    }

    private fun tryComputeHeadingFromFallback(timestamp: Long) {
        if (!hasAccelerometer || !hasMagnetometer) return

        // Throttle updates to 20Hz
        if (shouldThrottleUpdate(timestamp)) return

        // Compute rotation matrix from accelerometer and magnetometer
        val success = SensorManager.getRotationMatrix(
            rotationMatrix,
            null,
            accelerometerReading,
            magnetometerReading
        )

        if (!success) {
            _headingConfidence.value = HeadingConfidence.WEAK
            return
        }

        // Get orientation angles
        SensorManager.getOrientation(rotationMatrix, orientationAngles)

        // Convert azimuth from radians to degrees (0-360)
        val azimuthRadians = orientationAngles[0]
        var azimuthDegrees = Math.toDegrees(azimuthRadians.toDouble()).toFloat()

        // Normalize to 0-360 range
        if (azimuthDegrees < 0) {
            azimuthDegrees += 360f
        }

        updateHeading(azimuthDegrees)
    }

    private fun updateHeading(heading: Float) {
        _currentHeading.value = heading

        // If we don't have explicit accuracy info, assume good if we're getting updates
        if (_headingConfidence.value == HeadingConfidence.UNAVAILABLE) {
            _headingConfidence.value = HeadingConfidence.GOOD
        }

        Log.d(TAG, "Heading updated: $heading° (confidence: ${_headingConfidence.value})")

        // Reset timeout since we're getting updates
        if (isUpdating) {
            resetTimeoutTimer()
        }
    }

    private fun shouldThrottleUpdate(timestamp: Long): Boolean {
        if (lastUpdateTimeNs == 0L) {
            lastUpdateTimeNs = timestamp
            return false
        }

        val elapsed = timestamp - lastUpdateTimeNs
        if (elapsed >= MIN_UPDATE_INTERVAL_NS) {
            lastUpdateTimeNs = timestamp
            return false
        }

        return true
    }

    private fun resetTimeoutTimer() {
        cancelTimeoutTimer()

        timeoutTimer = Timer().apply {
            schedule(object : TimerTask() {
                override fun run() {
                    Log.w(TAG, "Heading updates timed out after ${TIMEOUT_DURATION_MS}ms")
                    stopUpdates()
                }
            }, TIMEOUT_DURATION_MS)
        }
    }

    private fun cancelTimeoutTimer() {
        timeoutTimer?.cancel()
        timeoutTimer = null
    }
}
